File: app/src/main/java/com/catsmoker/app/RootActivity.java
package com.catsmoker.app;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.topjohnwu.superuser.Shell;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class RootActivity extends AppCompatActivity {

    // This field will be set to true by the Xposed module if it's active.
    public static boolean isModuleActive = false;

    private enum LsposedStatus {
        NOT_ACTIVE, // Module is not active
        ACTIVE      // Module is active
    }

    private TextView statusRootText;
    private TextView tvLsposedStatus;
    private MaterialButton btnRefresh;
    private View rootView;

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_root);

        setupToolbar();
        initViews();
        setupListeners();

        refreshStatus();
    }

    private void setupToolbar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(R.string.root_status_title);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    private void initViews() {
        rootView = findViewById(android.R.id.content);
        statusRootText = findViewById(R.id.tv_root_status);
        tvLsposedStatus = findViewById(R.id.tv_lsposed_status);
        btnRefresh = findViewById(R.id.btn_refresh);
    }

    private void setupListeners() {
        btnRefresh.setOnClickListener(v -> refreshStatus());

        View btnInstall = findViewById(R.id.btn_install_lsposed);
        if (btnInstall != null) {
            btnInstall.setOnClickListener(v -> openUrl());
        }

        findViewById(R.id.btn_open_manager).setOnClickListener(v -> launchRootManager());
    }

    /**
     * Attempts to find and launch Magisk, KernelSU, or APatch.
     */
    private void launchRootManager() {
        Intent intent = getRootManagerIntent();
        if (intent != null) {
            startActivity(intent);
        } else {
            showSnackbar(getString(R.string.root_manager_not_found));
        }
    }

    private Intent getRootManagerIntent() {
        PackageManager pm = getPackageManager();
        String[] packages = {
                "com.topjohnwu.magisk", // Magisk
                "me.weishu.kernelsu",   // KernelSU
                "me.bmax.apatch"        // APatch
        };

        for (String pkg : packages) {
            Intent intent = pm.getLaunchIntentForPackage(pkg);
            if (intent != null) return intent;
        }
        return null;
    }

    private void refreshStatus() {
        btnRefresh.setEnabled(false);
        btnRefresh.setText(R.string.status_checking);

        statusRootText.setText(R.string.root_access_checking);
        statusRootText.setTextColor(ContextCompat.getColor(this, android.R.color.darker_gray));
        statusRootText.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
        
        tvLsposedStatus.setText(R.string.checking_lsposed);
        tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.darker_gray));
        tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);

        executor.execute(() -> {
            // FIX: 'rootAccess()' is deprecated.
            // Use Shell.getShell().isRoot() instead.
            // This acquires a shell (if needed) and checks uid == 0.
            boolean isRooted;
            try {
                isRooted = Shell.getShell().isRoot();
            } catch (Exception e) {
                isRooted = false;
            }

            // Check LSPosed Module status
            LsposedStatus lsposedModuleStatus = getLsposedModuleStatus();

            final boolean finalIsRooted = isRooted;
            mainHandler.post(() -> {
                if (isFinishing() || isDestroyed()) return; // Safety check

                updateUi(finalIsRooted, lsposedModuleStatus);
                btnRefresh.setEnabled(true);
                btnRefresh.setText(R.string.refresh_status);
                showSnackbar(getString(R.string.status_refreshed));
            });
        });
    }

    private void updateUi(boolean isRooted, LsposedStatus lsposedModuleStatus) {
        // Update Root Status UI
        if (isRooted) {
            statusRootText.setText(R.string.root_access_granted);
            statusRootText.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark));
            statusRootText.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_check_circle, 0, 0, 0);
        } else {
            statusRootText.setText(R.string.root_access_denied);
            statusRootText.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark));
            statusRootText.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_error, 0, 0, 0);
        }

        // Update LSPosed Status UI
        if (lsposedModuleStatus == LsposedStatus.ACTIVE) {
            tvLsposedStatus.setText(R.string.lsposed_module_enabled);
            tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark));
            tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_check_circle, 0, 0, 0);
        } else {
            tvLsposedStatus.setText(R.string.lsposed_module_disabled);
            tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark));
            tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_error, 0, 0, 0);
        }
    }

    private LsposedStatus getLsposedModuleStatus() {
        return isModuleActive ? LsposedStatus.ACTIVE : LsposedStatus.NOT_ACTIVE;
    }

    private void openUrl() {
        try {
            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/LSPosed/LSPosed/releases"));
            startActivity(browserIntent);
        } catch (Exception e) {
            showSnackbar(getString(R.string.could_not_open_browser));
        }
    }

    private void showSnackbar(String message) {
        if (rootView != null) {
            Snackbar.make(rootView, message, Snackbar.LENGTH_SHORT).show();
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        getOnBackPressedDispatcher().onBackPressed();
        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (executor != null && !executor.isShutdown()) {
            executor.shutdownNow();
        }
    }
}

File: app/src/main/java/com/catsmoker/app/RootActivity.kt
package com.catsmoker.app

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import androidx.lifecycle.lifecycleScope
import com.catsmoker.app.databinding.ActivityRootBinding
import com.google.android.material.snackbar.Snackbar
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class RootActivity : AppCompatActivity() {

    private lateinit var binding: ActivityRootBinding
    private val prefs by lazy {
        try {
            getSharedPreferences(LSPosedConfig.PREFS_NAME, MODE_WORLD_READABLE)
        } catch (_: SecurityException) {
            getSharedPreferences(LSPosedConfig.PREFS_NAME, MODE_PRIVATE)
        }
    }

    private enum class LsposedStatus {
        NOT_ACTIVE,  // Module is not active
        ACTIVE       // Module is active
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityRootBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupToolbar()
        setupListeners()
        bindLsposedConfig()
        refreshStatus()
    }

    private fun setupToolbar() {
        supportActionBar?.apply {
            setTitle(R.string.root_status_title)
            setDisplayHomeAsUpEnabled(true)
        }
    }

    private fun setupListeners() {
        binding.btnRefresh.setOnClickListener { refreshStatus() }
        binding.btnInstallLsposed.setOnClickListener { openUrl() }
        binding.btnOpenManager.setOnClickListener { launchRootManager() }
        binding.btnSaveLsposedConfig.setOnClickListener { saveLsposedConfig() }
    }

    private fun bindLsposedConfig() {
        val defaultTargets = LSPosedConfig.DEFAULT_TARGET_PACKAGES.joinToString("\n")
        val defaultProps = LSPosedConfig.DEFAULT_DEVICE_PROPS.entries.joinToString("\n") {
            "${it.key}=${it.value}"
        }

        binding.switchLsposedEnabled.setOnCheckedChangeListener(null)
        binding.switchLsposedEnabled.isChecked =
            prefs.getBoolean(LSPosedConfig.KEY_ENABLED, true)
        binding.etTargetPackages.setText(
            prefs.getString(LSPosedConfig.KEY_TARGET_PACKAGES, defaultTargets)
        )
        binding.etDeviceProps.setText(
            prefs.getString(LSPosedConfig.KEY_DEVICE_PROPS, defaultProps)
        )

        binding.switchLsposedEnabled.setOnCheckedChangeListener { _, isChecked ->
            prefs.edit().putBoolean(LSPosedConfig.KEY_ENABLED, isChecked).commit()
            showSnackbar(getString(R.string.lsposed_config_hint))
        }
    }

    private fun saveLsposedConfig() {
        val targetRaw = binding.etTargetPackages.text?.toString().orEmpty()
        val propsRaw = binding.etDeviceProps.text?.toString().orEmpty()

        val targets = parseTargetPackages(targetRaw)
        val props = parseDeviceProps(propsRaw)

        if (targets.isEmpty() || props.isEmpty()) {
            showSnackbar(getString(R.string.lsposed_config_invalid))
            return
        }

        val normalizedTargets = TextUtils.join("\n", targets)
        val normalizedProps = props.entries.joinToString("\n") { "${it.key}=${it.value}" }

        prefs.edit()
            .putString(LSPosedConfig.KEY_TARGET_PACKAGES, normalizedTargets)
            .putString(LSPosedConfig.KEY_DEVICE_PROPS, normalizedProps)
            .commit()

        binding.etTargetPackages.setText(normalizedTargets)
        binding.etDeviceProps.setText(normalizedProps)

        showSnackbar(
            "${getString(R.string.lsposed_config_saved)} ${getString(R.string.lsposed_config_hint)}"
        )
    }

    private fun parseTargetPackages(raw: String): Set<String> {
        if (raw.isBlank()) return emptySet()
        val result = LinkedHashSet<String>()
        raw
            .split("\n", ",")
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .forEach { result.add(it) }
        return result
    }

    private fun parseDeviceProps(raw: String): Map<String, String> {
        if (raw.isBlank()) return emptyMap()
        val result = LinkedHashMap<String, String>()
        raw.lineSequence().forEach { line ->
            val trimmed = line.trim()
            if (trimmed.isEmpty()) return@forEach
            val idx = trimmed.indexOf('=')
            if (idx <= 0 || idx == trimmed.length - 1) return@forEach
            val key = trimmed.substring(0, idx).trim()
            val value = trimmed.substring(idx + 1).trim()
            if (key.isNotEmpty() && value.isNotEmpty()) {
                result[key] = value
            }
        }
        return result
    }


    /**
     * Attempts to find and launch Magisk, KernelSU, or APatch.
     */
    private fun launchRootManager() {
        val intent = rootManagerIntent
        if (intent != null) {
            startActivity(intent)
        } else {
            showSnackbar(getString(R.string.root_manager_not_found))
        }
    }

    private val rootManagerIntent: Intent?
        get() {
            val pm = packageManager
            val packages = arrayOf(
                "com.topjohnwu.magisk",  // Magisk
                "me.weishu.kernelsu",    // KernelSU
                "me.bmax.apatch"         // APatch
            )

            for (pkg in packages) {
                val intent = pm.getLaunchIntentForPackage(pkg)
                if (intent != null) return intent
            }
            return null
        }

    private fun refreshStatus() {
        binding.btnRefresh.isEnabled = false
        binding.btnRefresh.setText(R.string.status_checking)

        binding.tvRootStatus.setText(R.string.root_access_checking)
        binding.tvRootStatus.setTextColor(ContextCompat.getColor(this, android.R.color.darker_gray))
        binding.tvRootStatus.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)

        binding.tvLsposedStatus.setText(R.string.checking_lsposed)
        binding.tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.darker_gray))
        binding.tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)

        lifecycleScope.launch(Dispatchers.IO) {
            // Check Root Access
            val isRooted = try {
                Shell.getShell().isRoot
            } catch (_: Exception) {
                false
            }

            // Check LSPosed Module status
            val lsposedModuleStatus = if (isModuleActive) LsposedStatus.ACTIVE else LsposedStatus.NOT_ACTIVE

            withContext(Dispatchers.Main) {
                if (isFinishing || isDestroyed) return@withContext

                updateUi(isRooted, lsposedModuleStatus)
                binding.btnRefresh.isEnabled = true
                binding.btnRefresh.setText(R.string.refresh_status)
                showSnackbar(getString(R.string.status_refreshed))
            }
        }
    }

    private fun updateUi(isRooted: Boolean, lsposedModuleStatus: LsposedStatus) {
        // Update Root Status UI
        if (isRooted) {
            binding.tvRootStatus.setText(R.string.root_access_granted)
            binding.tvRootStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark))
            binding.tvRootStatus.setCompoundDrawablesWithIntrinsicBounds(
                android.R.drawable.checkbox_on_background, 0, 0, 0
            )
        } else {
            binding.tvRootStatus.setText(R.string.root_access_denied)
            binding.tvRootStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
            binding.tvRootStatus.setCompoundDrawablesWithIntrinsicBounds(android.R.drawable.ic_dialog_alert, 0, 0, 0)
        }

        // Update LSPosed Status UI
        if (lsposedModuleStatus == LsposedStatus.ACTIVE) {
            binding.tvLsposedStatus.setText(R.string.lsposed_module_enabled)
            binding.tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark))
            binding.tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(
                android.R.drawable.checkbox_on_background, 0, 0, 0
            )
        } else {
            binding.tvLsposedStatus.setText(R.string.lsposed_module_disabled)
            binding.tvLsposedStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
            binding.tvLsposedStatus.setCompoundDrawablesWithIntrinsicBounds(android.R.drawable.ic_dialog_alert, 0, 0, 0)
        }
    }

    private fun openUrl() {
        try {
            val browserIntent = Intent(Intent.ACTION_VIEW, "https://github.com/LSPosed/LSPosed/releases".toUri())
            startActivity(browserIntent)
        } catch (_: Exception) {
            showSnackbar(getString(R.string.could_not_open_browser))
        }
    }

    private fun showSnackbar(message: String) {
        Snackbar.make(binding.root, message, Snackbar.LENGTH_SHORT).show()
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressedDispatcher.onBackPressed()
        return true
    }

    companion object {
        // This field will be set to true by the Xposed module if it's active.
        @JvmField
        var isModuleActive: Boolean = false
    }
}
